logo-ann/generation/02_generate_embeddings.py

```diff
@@ -39,46 +64,101 @@ def generate_embeddings_iter(
+    - file_path: path of the hdf5 file containing the data of all the logos
+    - batch-size: size of each batche of logos embedded at the same time 
+    - device: hardware used to compute the embeddings
+    - seen_set: set of every logo already embedded in 
+    - min-confidence: minimum of confidence allowed for a logo to be accepted as one
+
+    Yield the following outputs:
+    - embeddings: embeddings of every logo of the yielded batch 
+    - external_id: id of the logo
+    """
+
     with h5py.File(str(file_path), "r") as f:
         image_dset = f["image"]
         confidence_dset = f["confidence"]
         external_id_dset = f["external_id"]
 
         for slicing in chunked(range(len(image_dset)), batch_size):
-            slicing = np.array(slicing)
+            slicing = np.array(
+                slicing
+            )  # slicing was a list, we make an array out of it
             external_ids = external_id_dset[slicing]
             mask = external_ids == 0
 
-            if np.all(mask):
+            if np.all(
+                mask
+            ):  # if we only have zeros as external ideas, it's an empty batch and it means we won't have any other logos so we stop the loop
                 break
 
-            mask = (~mask) & (confidence_dset[slicing] >= min_confidence)
+            mask = (~mask) & (
+                confidence_dset[slicing] >= min_confidence
+            )  # we add to the mask the confidence parameter to avoid embedding
 
             for i, external_id in enumerate(external_ids):
                 if int(external_id) in seen_set:
                     mask[i] = 0
 
-            if np.all(~mask):
+            if np.all(~mask):  # if we only have
                 continue
 
             images = image_dset[slicing][mask]
             images = np.moveaxis(images, -1, 1)  # move channel dim to 1st dim
 
+            """### If using efficientnet models :
             with torch.no_grad():
                 torch_images = torch.tensor(images, dtype=torch.float32, device=device)
                 embeddings = model.extract_features(torch_images).cpu().numpy()
+            
 
             max_embeddings = np.max(embeddings, (-1, -2))
             yield (
                 max_embeddings,
                 external_ids[mask],
             )
+            ###
+            """
+
+            ### If using CLIP models :
+            with torch.no_grad():
+                array_to_PIL = lambda x: PIL.Image.fromarray(
+                    x, mode="RGB"
+                )  # convert the np.array to PIL in order to use the CLIProcessor
+                images = processor(
+                    images=[array_to_PIL(images[i]) for i in range(batch_size)],
+                    return_tensors="pt",
+                )[
+                    "pixel_values"
+                ]  # preprocess the images to put them into the model
+                embeddings = (
+                    model(**{"pixel_values": images.to(device)})
```

Comment:
more straighforward to do `model(pixel_values=images.to(device))`

Reply:
Ok. I must admit I just saw that you had done this in the embedding model benchmark and I followed. I don't understand this syntax really well. What do the **{...} do exactly ?